packages/zoe/test/unitTests/contracts/test-operaConcertTicket.js

```diff
@@ -0,0 +1,409 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { test } from 'tape-promise/tape';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import bundleSource from '@agoric/bundle-source';
+import harden from '@agoric/harden';
+import produceIssuer from '@agoric/ertp';
+import { makeZoe } from '../../../src/zoe';
+
+const operaConcertTicketRoot = `${__dirname}/../../../src/contracts/operaConcertTicket`;
+
+test(`__Test Scenario__
+
+The Opera de Bordeaux plays the contract creator and the auditorium
+It creates the contract for a show ("Steven Universe, the Opera", Web, March 25th 2020 at 8pm, 3 tickets)
+The Opera wants 22 moolas per ticket
+
+Alice buys ticket #1
+
+Bob tries to buy ticket 1 and fails. He buys ticket #2 and #3
+
+The Opera is told about the show being sold out. It gets all the moolas from the sale`, async t => {
+  // Setup initial conditions
+  const {
+    mint: moolaMint,
+    issuer: moolaIssuer,
+    amountMath: { make: moola },
+  } = produceIssuer('moola');
+
+  const zoe = makeZoe({ require });
+  const inviteIssuer = zoe.getInviteIssuer();
+
+  // === Initial Opera de Bordeaux part ===
+  const contractReadyP = bundleSource(operaConcertTicketRoot).then(
+    ({ source, moduleFormat }) => {
+      const expectedAmountPerTicket = moola(22);
+
+      const installationHandle = zoe.install(source, moduleFormat);
+
+      return zoe
+        .makeInstance(installationHandle, harden({ Money: moolaIssuer }), {
+          show: 'Steven Universe, the Opera',
+          start: 'Web, March 25th 2020 at 8pm',
+          count: 3,
+          expectedAmountPerTicket,
+        })
+        .then(auditoriumInvite => {
+          return inviteIssuer
+            .getAmountOf(auditoriumInvite)
+            .then(
+              ({ extent: [{ instanceHandle: auditoriumInstanceHandle }] }) => {
+                const { publicAPI } = zoe.getInstance(auditoriumInstanceHandle);
+
+                t.equal(
+                  typeof publicAPI.makeBuyerInvite,
+                  'function',
+                  'publicAPI.makeBuyerInvite should be a function',
+                );
+                t.equal(
+                  typeof publicAPI.getTicketIssuer,
+                  'function',
+                  'publicAPI.getTicketIssuer should be a function',
+                );
+                t.equal(
+                  typeof publicAPI.getAvailableTickets,
+                  'function',
+                  'publicAPI.getAvailableTickets should be a function',
+                );
+
+                // The auditorium redeems its invite.
+                return (
+                  zoe
+                    .redeem(auditoriumInvite, harden({}))
```

Comment:
oh... I'm realizing now that since the auditorium `want`s nothing, the buyers can get the tickets for free.
This is what we discussed yesterday and i wonder now whether that's a good idea. It sorts of offers zero guarantees for the auditorium since it's unable to express what hey want

Eventually, they'll be able to express the proportionality thing, but right now, it'd be good to have some level of guarantee as well

Reply:
Ah, sorry, I missed this comment and replied in the google doc. We can add a check to `performExchange` such that if the buyer is offering less than the ticket price, we reject the offer immediately. Further enforcement of offer safety, I think we can add in another PR or another tutorial.

Reply:
> We can add a check to performExchange such that if the buyer is offering less than the ticket price, we reject the offer immediately.

I can do that, but I feel like this would be reimplementing what zoe is supposed to do for us

and that's what gives me the feeling that this contract is not a good example of all the safety Zoe could offer: as it is now, Zoe guarantees offer safety for the buyer, but not the auditorium. With your suggestion, Zoe+contract guarantees offer safety... but Zoe is supposed to be protecting against contract code malfunctions. I feel like the auditorium should be getting offer safety from Zoe alone, like the buyer gets

I'll wait for @erights review and procede with your suggestion if there's an agreement

Reply:
Thanks for waiting and sorry my feedback took so long. As mentioned above, we do have a plan for Zoe enforcement of offer safety for this case. The so-called "multiples" offer I explain above. That's not going to happen soon. Until then, the auditorium's meaningful offer safety depends on the contract, not on Zoe.

Even after we add multiples, exclusive choice, offsets, and perhaps other things we find we need, there will always be an expressiveness limit to the offer safety Zoe itself can guarantee. The case that will probably be most painful soonest is offers in terms of "current" market prices, such as a "market order" entered into an exchange. We have no plans to have Zoe enforce safety relative to dynamic market prices at the time. This would always have to be delegated up to the contract.

Cases like market prices, which will come up a lot across many contracts, might cause another partitioning of risk within what Zoe regards as the contract level. @katelynsills did some good investigation of this early on, of generically parameterizable contract templates, that specialized code could be plugged into to create custom contracts. The templates could be highly vetted highly reusable frameworks. Such a template might be the code enforcing such higher-level safety properties, independent of what the specialized code is doing. This applies the Zoe risk partitioning principle at a higher layer of abstraction. The key thing is that multiple such templates can co-exist, each of which understand a different "language" for expressing such higher level offer safety constraints. No one such language need be universal.

I look forward to the day we get back to exploring that.